### PR TITLE
Recursor: Load IPv6 entries from `etc-hosts` file

### LINKED ISSUE #2248: Recursor - Not loading IPv6 entries from etc-hosts-file.
The use case is simple.  I need to inject a couple of RRs internally to a zone that I am hosting externally. 

I created a new etc-hosts-file.  Added two entries in standard format.  One for v4, one for v6.  Enabled the export and restarted.  "host host.example.com" only returns the v4 address and "dig aaaa host.example.com" returns no v6 address.

Expected behavior is that the v6 address is returned on query as well.

### PR DESCRIPTION
**Pull request by fredmorcos**
### Short description
This adds the missing parts to load IPv6 entries from a provided `etc-hosts` file:
* Adds support for reversing IPv4 and IPv6 addresses to `ComboAddress`. Also adds tests for that.
* Replaces the usage of strings for IP addresses with `ComboAddress`.
* Moves a few things around for testability (in a separate commit).
* Also breaks down some parts into smaller functions (also in a separate commit).
* Fixes a pre-existing issue where `PTR` entries were being added for aliases (i.e. "non canonical hostname(s)").
* Adds tests for the pre-existing IPv4-only functionality in `test-reczones-helpers.cc`.
* Also adds tests for loading IPv6 entries as well.
* Adds the ability to print domain maps and their corresponding record entries to a `std::stringstream`.
* The usual misc cleanups & clang-tidy fixes.

Closes #2248 

PS: To run the tests with print output use `./testrunner --run_test=... --log_level=message`

### Checklist
I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [X] added or modified regression test(s)
- [X] added or modified unit test(s)

### CODE REVIEW

File: pdns/reczones.cc [hunk 1]
```diff
@@ -197,7 +197,7 @@ void primeRootNSZones(DNSSECMode mode, unsigned int depth)
   }
 }
 
-static void makeNameToIPZone(std::shared_ptr<SyncRes::domainmap_t> newMap, const DNSName& hostname, const string& ip, Logr::log_t log)
+static void makeNameToIPZone(const std::shared_ptr<SyncRes::domainmap_t>& newMap, const DNSName& hostname, const string& ip, Logr::log_t log)
```
Comment (cmouse) [inline pdns/reczones.cc]: is this some lint that is in CI pipeline too?

File: pdns/reczones.cc [hunk 2]
```diff
@@ -237,15 +237,15 @@ static void makeNameToIPZone(const std::shared_ptr<SyncRes::domainmap_t>& newMap
...
+  vector<string> ipParts;
+  stringtok(ipParts, address, ".");
 
   SyncRes::AuthDomain ad;
   ad.d_rdForward = false;
 
   DNSRecord dr;
-  for (int n = ipparts.size() - 1; n >= 0; --n) {
-    dr.d_name.appendRawLabel(ipparts[n]);
+  for (auto part = ipParts.rbegin(); part != ipParts.rend(); ++part) {
```
Comment (cmouse) [inline pdns/reczones.cc]: `const auto& part`

File: ext/lmdb-safe/lmdb-safe.hh [hunk 3]
```diff
@@ -1,4 +1,5 @@
 #pragma once
+#include <string_view>
```
Comment (cmouse) [inline ext/lmdb-safe/lmdb-safe.hh:2]: this commit seems totally unrelated here.

File: pdns/dnsparser.hh [hunk 4]
```diff
@@ -304,6 +305,18 @@ struct DNSRecord
   uint16_t d_clen{};
   DNSResourceRecord::Place d_place;
 
+  [[nodiscard]] std::string toString(const std::string& indent = "") const
```
Comment (cmouse) [inline pdns/dnsparser.hh:308]: this does not look like toString for a record, more like some special represetantion.

File: pdns/reczones-helpers.cc [hunk 5]
```diff
@@ -27,115 +27,129 @@
...
                                        Logr::log_t log)
 {
-  for (unsigned int n = 1; n < parts.size(); ++n) {
-    if (searchSuffix.empty() || parts[n].find('.') != string::npos) {
-      makeNameToIPZone(newMap, DNSName(parts[n]), parts[0], log);
+  const ComboAddress address{parts[0]};
+
+  // Go over the hostname and aliases (parts[1], parts[2], etc...) and add entries
+  // for forward lookups.
+  for (auto name = parts.cbegin() + 1; name != parts.cend(); ++name) {
```
Comment (cmouse) [inline pdns/reczones-helpers.cc:136]: `const auto& name`

**Review by cmouse — CHANGES_REQUESTED** [review]

File: pdns/reczones.cc (hunk 1 above)
Comment (fredmorcos) [reply pdns/reczones.cc]: Not yet, but I want to get it integrated.

File: ext/lmdb-safe/lmdb-safe.hh (hunk 3 above)
Comment (fredmorcos) [reply ext/lmdb-safe/lmdb-safe.hh:2]: It was to fix an error message I was getting in my IDE, so it's part of "cleanup". Is it OK if I keep it?

File: pdns/iputils.hh [hunk 6]
```diff
@@ -283,6 +283,32 @@ union ComboAddress {
...
+      auto c = (ip >> 16) & 0xFF;
+      auto d = (ip >> 24) & 0xFF;
+      return std::to_string(a) + "." + std::to_string(b) + "." + std::to_string(c) + "." + std::to_string(d);
+    }
+    else {
+      const auto* addr = &sin6.sin6_addr;
+      std::stringstream res{};
+      res << std::hex;
+      for (int i = 15; i >= 0; i--) {
+        auto byte = (int)addr->s6_addr[i];
```
Comment (cmouse) [inline pdns/iputils.hh:301]: why this (int) cast here?

File: pdns/reczones-helpers.cc (hunk 5 above)
Comment (fredmorcos) [reply pdns/reczones-helpers.cc:136]: I must be missing something here. Doesn't the `const` in your suggestion apply to the iterator object here? How would the iterator advance if it cannot be mutated?

File: pdns/reczones.cc (hunk 2 above)
Comment (fredmorcos) [reply pdns/reczones.cc]: This was removed later in favor of `ComboAddress`.

File: pdns/reczones-helpers.cc (hunk 5 above)
Comment (cmouse) [reply pdns/reczones-helpers.cc:136]: oh right it's iterator... my bad.

File: ext/lmdb-safe/lmdb-safe.hh (hunk 3 above)
Comment (Habbie) [reply ext/lmdb-safe/lmdb-safe.hh:2]: this can stay

File: pdns/iputils.hh (hunk 6 above)
Comment (fredmorcos) [reply pdns/iputils.hh:301]: Probably a left-over.

File: pdns/dnsparser.hh (hunk 4 above)
Comment (fredmorcos) [reply pdns/dnsparser.hh:308]: You mean it should be renamed to something else?

Comment (cmouse) [reply pdns/dnsparser.hh:308]: yes.

**Comment by omoerbeek** [comment]
I did not look at the code yet, but did a few tests.

Looking at at an example `/etc/hosts`  containing
```
194.109.6.2     p
194.109.6.3     p
2001:888:0:7::2 p
2001:888:0:7::2 q
```
I see that `dig -t A p` succeeds but has one record. `dig -t AAAA p` gives me an empty answer section. `dig -t AAAA q` works and gives me a single record.

In an ideal world the forward mapping would work for all cases: two answer records for `dig -t A p`, one record for `dig -t AAAA p` and one record for `dig -t AAAA q`.

Having both an IPv4 and an IPv6 address for the same name *is* very common. So it would be nice if we would handle at least that case.

I don't know if this is feasible with the current structure, I guess it would mean merging domains in some cases. 

The reverse (PTR) mapping should give only the first name encountered for each address. This seems to work as expected.

**Comment by omoerbeek** [comment]
To  be clear, the behavior above is already present in master, so it is not a regression.

File: pdns/dnsparser.hh [hunk 7]
```diff
@@ -289,17 +290,33 @@ protected:
...
+  DNSRecord(const std::string& name, std::shared_ptr<DNSRecordContent> content, const uint16_t type, const uint16_t qclass = QClass::IN, const uint32_t ttl = 86400, const uint16_t clen = 0, const DNSResourceRecord::Place place = DNSResourceRecord::ANSWER) :
+    d_name(DNSName(name)), d_content(std::move(content)), d_type(type), d_class(qclass), d_ttl(ttl), d_clen(clen), d_place(place) {}
+
   DNSName d_name;
   std::shared_ptr<DNSRecordContent> d_content;
   uint16_t d_type;
-  uint16_t d_class;
-  uint32_t d_ttl;
-  uint16_t d_clen;
+  uint16_t d_class{};
```
Comment (omoerbeek) [inline pdns/dnsparser.hh:303]: Why this empty initialization? Isn't it the default? Or are you trying to make that explicit? In that case it make sense.

File: pdns/syncres.hh [hunk 8]
```diff
@@ -114,6 +114,38 @@ public:
...
+    bool operator==(const AuthDomain& rhs) const
+    {
+      return
+        d_records == rhs.d_records &&
+        d_servers == rhs.d_servers &&
+        d_name == rhs.d_name &&
+        d_rdForward == rhs.d_rdForward;
+    }
+
+    [[nodiscard]] std::string toString(const std::string& indent = "", const std::string& indentLevel = "  ") const
```
Comment (omoerbeek) [inline pdns/syncres.hh:126]: I think this is also not a generic toString(), more of a debug aid, so a better fitting name is needed.

File: pdns/reczones.cc [hunk 9]
```diff
@@ -197,100 +200,14 @@ void primeRootNSZones(DNSSECMode mode, unsigned int depth)
...
-
 static void convertServersForAD(const std::string& zone, const std::string& input, SyncRes::AuthDomain& ad, const char* sepa, Logr::log_t log, bool verbose = true)
 {
   vector<string> servers;
   stringtok(servers, input, sepa);
   ad.d_servers.clear();
 
   vector<string> addresses;
-  for (vector<string>::const_iterator iter = servers.begin(); iter != servers.end(); ++iter) {
+  for (auto iter = servers.begin(); iter != servers.end(); ++iter) {
```
Comment (omoerbeek) [inline pdns/reczones.cc:210]:
```suggestion
  for (const auto& server : servers) {
```

File: pdns/reczones.cc [hunk 10]
```diff
@@ -197,100 +200,14 @@ void primeRootNSZones(DNSSECMode mode, unsigned int depth)
...
     ComboAddress addr = parseIPAndPort(*iter, 53);
```
Comment (omoerbeek) [inline pdns/reczones.cc:211]:
```suggestion
    ComboAddress addr = parseIPAndPort(server, 53);
```

File: pdns/dnsparser.hh (hunk 7 above)
Comment (fredmorcos) [reply pdns/dnsparser.hh:303]: It is to be explicit, and then there's no need to initialize them in the constructors. I believe without assignment, members don't get "zero values" and instead just get pre-existing memory garbage.

**Comment by omoerbeek** [comment]
> @omoerbeek I've opened an issue regarding the problem you mention: #11690 I think we could get this PR merged since it doesn't introduce a regression, and I could work on the issue as a separate PR. I hope that is alright with you.

Great! I intend to finish the code review soon.

File: pdns/dnsparser.hh [hunk 11]
```diff
@@ -291,19 +291,19 @@ protected:
 struct DNSRecord
 {
   DNSRecord() :
-    d_type(0), d_class(QClass::IN), d_place(DNSResourceRecord::ANSWER)
```
Comment (omoerbeek) [inline pdns/dnsparser.hh]: In the original, type was always initialized (either to zero or to the argument to the ct). The default constructor now only inits `d_class` (explicitly) and `d_place` (implicitly), leaving `d_type` uninitialized. I'm worried there is code that depends on the zero value.
The original and the PR leave `d_ttl` and `d_clen` uninitialized in the default ct, but we can be almost be sure nothing depends on that. For `d_type` that is not clear.

File: pdns/syncres.hh [hunk 12]
```diff
@@ -114,6 +114,38 @@ public:
     DNSName d_name;
     bool d_rdForward{false};
 
+    bool operator==(const AuthDomain& rhs) const
```
Comment (omoerbeek) [inline pdns/syncres.hh:117]: IMO both these functions do not need to be inline. The code is only used for debugging and this header file is already too big...

File: pdns/dnsparser.hh [hunk 13]
```diff
@@ -289,16 +290,44 @@ protected:
...
-  uint32_t d_ttl;
-  uint16_t d_clen;
-  DNSResourceRecord::Place d_place;
+  uint16_t d_type{};
+  uint16_t d_class{};
+  uint32_t d_ttl{};
+  uint16_t d_clen{};
+  DNSResourceRecord::Place d_place = DNSResourceRecord::ANSWER;
+
+  [[nodiscard]] std::string print(const std::string& indent = "") const
```
Comment (omoerbeek) [inline pdns/dnsparser.hh:320]: No need for inlining here

**Review by omoerbeek — CHANGES_REQUESTED** [review]

File: pdns/dnsparser.hh (hunk 11 above)
Comment (fredmorcos) [reply pdns/dnsparser.hh]: `d_type` is being initialized to 0 because of `uint16_t d_type{};` and the PR does not leave `d_ttl` and `d_clen` uninitialized because of:
```
uint32_t d_ttl{};
uint16_t d_clen{};
```

Am I missing something?

Comment (omoerbeek) [reply pdns/dnsparser.hh]: Nope, I was missing something, I read the `{}` as default init, while it is actually zero-init.

Comment (fredmorcos) [reply pdns/dnsparser.hh]: Umm.. what is the difference?

Comment (omoerbeek) [reply pdns/dnsparser.hh]: For integral fields the default (absent) initializer means not initialized at all, i.e. garbage. For some reason in equated `{}` as the default initialized in my initial comment, while it actually does init to zero.

Comment (fredmorcos) [reply pdns/dnsparser.hh]: Right. Thanks for the clarification.

**Review by omoerbeek — APPROVED** [review]
Should be good to merge after conflict is resolved.
